### PR TITLE
Port the compiler gym service backends to the new API

### PR DESCRIPTION
**Pull request by ChrisCummins**
Issue #254.

### CODE REVIEW

File: examples/example_compiler_gym_service/service_cc/ExampleService.cc
Comment (hughleat) [inline examples/example_compiler_gym_service/service_cc/ExampleService.cc:13]: Unnecessary use of R string?

File: examples/example_compiler_gym_service/service_cc/ExampleService.cc
Comment (hughleat) [inline examples/example_compiler_gym_service/service_cc/ExampleService.cc:37]: Shame these are protobufs, then they could just be params to the constructor.

File: examples/example_compiler_gym_service/env_tests.py
Comment (hughleat) [inline examples/example_compiler_gym_service/env_tests.py:40]: Is this function unused?

**Review by hughleat — APPROVED** [review]
LGTM, although I still think that not making the user know anything about what RPC mechanism you are using would be better, including allowing a simpler API for them.

File: examples/example_compiler_gym_service/service_cc/ExampleService.cc
Comment (ChrisCummins) [reply examples/example_compiler_gym_service/service_cc/ExampleService.cc:13]: 👍

File: examples/example_compiler_gym_service/service_cc/ExampleService.cc
Comment (ChrisCummins) [reply examples/example_compiler_gym_service/service_cc/ExampleService.cc:37]: The problem is that this class is never instantiated by the user. They pass it into `CompilerGymService<T>` which handles lifetime management: 
https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/service/runtime/CompilerGymServiceImpl.h#L64

File: examples/example_compiler_gym_service/env_tests.py
Comment (ChrisCummins) [reply examples/example_compiler_gym_service/env_tests.py:40]: `@pytest.fixture` decorator declares a fixture, which means that any test function can then take `bin` as an argument and it will run and return the fixture value

**Comment by ChrisCummins** [comment]
Thanks for the review @hughleat!

The CI is complaining of a problem that I've never seen before. An undefined reference to the destructor of `google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()` and `google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()` in `getSpaces()`:

```
bazel-out/k8-fastbuild/bin/examples/_objs/RandomSearch/RandomSearch.pic.o:RandomSearch.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
```

Will investigate and fix before merging.

**Comment by ChrisCummins** [comment]
I have been blocked on this issue for a couple of days now and cannot seem to figure it out. I have been unable to reproduce this locally on any build environment, including the docker environment used for building the linux wheels. The problem affects all targets that use the new C++ CompilerGymService implementation, and produces the same link-time error:

```
ERROR: /home/runner/work/CompilerGym/CompilerGym/compiler_gym/envs/llvm/service/BUILD:53:10: Linking compiler_gym/envs/llvm/service/compiler_gym-llvm-service-prelinked failed: (Exit 1): clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/1570/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/compiler_gym-llvm-service-prelinked-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/1570/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/compiler_gym-llvm-service-prelinked-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/_objs/compiler_gym-llvm-service-prelinked/RunService.pic.o:RunService.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/_objs/compiler_gym-llvm-service-prelinked/RunService.pic.o:RunService.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/_objs/compiler_gym-llvm-service-prelinked/RunService.pic.o:RunService.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/compiler_gym/envs/llvm/service/_objs/compiler_gym-llvm-service-prelinked/RunService.pic.o:RunService.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

```
ERROR: /home/runner/work/CompilerGym/CompilerGym/examples/example_compiler_gym_service/service_cc/BUILD:7:10: Linking examples/example_compiler_gym_service/service_cc/compiler_gym-example-service-cc failed: (Exit 1): clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/2646/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/compiler_gym-example-service-cc-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/2646/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/compiler_gym-example-service-cc-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/_objs/compiler_gym-example-service-cc/ExampleService.pic.o:ExampleService.cc:function compiler_gym::runtime::CompilerGymService<(anonymous namespace)::ExampleCompilationSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/_objs/compiler_gym-example-service-cc/ExampleService.pic.o:ExampleService.cc:function compiler_gym::runtime::CompilerGymService<(anonymous namespace)::ExampleCompilationSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/_objs/compiler_gym-example-service-cc/ExampleService.pic.o:ExampleService.cc:function compiler_gym::runtime::CompilerGymService<(anonymous namespace)::ExampleCompilationSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/example_compiler_gym_service/service_cc/_objs/compiler_gym-example-service-cc/ExampleService.pic.o:ExampleService.cc:function compiler_gym::runtime::CompilerGymService<(anonymous namespace)::ExampleCompilationSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

```
ERROR: /home/runner/work/CompilerGym/CompilerGym/examples/BUILD:120:10: Linking examples/RandomSearch failed: (Exit 1): clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/2798/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/examples/RandomSearch-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox clang failed: error executing command 
  (cd /home/runner/.cache/bazel/_bazel_runner/db3e34f1e14ee53d927a48b4c425fcaa/sandbox/linux-sandbox/2798/execroot/CompilerGym && \
  exec env - \
    LD_LIBRARY_PATH=/opt/hostedtoolcache/Python/3.9.5/x64/lib \
    PATH=/home/runner/.cache/bazelisk/downloads/bazelbuild/bazel-4.0.0-linux-x86_64/bin:/opt/hostedtoolcache/Python/3.9.5/x64/bin:/opt/hostedtoolcache/Python/3.9.5/x64:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin \
    *** \
  /usr/bin/clang @bazel-out/k8-fastbuild/bin/examples/RandomSearch-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-fastbuild/bin/examples/_objs/RandomSearch/RandomSearch.pic.o:RandomSearch.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/_objs/RandomSearch/RandomSearch.pic.o:RandomSearch.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/_objs/RandomSearch/RandomSearch.pic.o:RandomSearch.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ActionSpace>::~RepeatedPtrField()'
bazel-out/k8-fastbuild/bin/examples/_objs/RandomSearch/RandomSearch.pic.o:RandomSearch.cc:function compiler_gym::runtime::CompilerGymService<compiler_gym::llvm_service::LlvmSession>::GetSpaces(grpc::ServerContext*, compiler_gym::GetSpacesRequest const*, compiler_gym::GetSpacesReply*): error: undefined reference to 'google::protobuf::RepeatedPtrField<compiler_gym::ObservationSpace>::~RepeatedPtrField()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by ChrisCummins** [comment]
Fixed the destructor linker error by rewriting the code using for loops rather than direct vector assignment. There's still a few bits to fix up in the CI configs then I'll merge.

**Comment by ChrisCummins** [comment]
All test signals green. Merging!

**Pull request merged**
